src/ansiblelint/transformer.py

```diff
@@ -62,17 +63,37 @@ def run(self) -> None:
                 data = ""
                 file_is_yaml = False
 
+            ruamel_data: Optional[Union[CommentedMap, CommentedSeq]] = None
             if file_is_yaml:
                 # We need a fresh YAML() instance for each load because ruamel.yaml
                 # stores intermediate state during load which could affect loading
                 # any other files. (Based on suggestion from ruamel.yaml author)
                 yaml = FormattedYAML()
 
-                ruamel_data: Union[CommentedMap, CommentedSeq] = yaml.loads(data)
+                ruamel_data = yaml.loads(data)
                 if not isinstance(ruamel_data, (CommentedMap, CommentedSeq)):
                     # This is an empty vars file or similar which loads as None.
                     # It is not safe to write this file or data-loss is likely.
                     # Only maps and sequences can preserve comments. Skip it.
                     continue
+
+            self._do_transforms(file, ruamel_data or data, matches)
+
+            if file_is_yaml:
+                # noinspection PyUnboundLocalVariable
                 file.content = yaml.dumps(ruamel_data)
+
+            if file.updated:
                 file.write()
+
+    @staticmethod
+    def _do_transforms(
+        file: Lintable,
+        data: Union[CommentedMap, CommentedSeq, str],
+        matches: List[MatchError],
+    ) -> None:
+        """Do Rule-Transforms handling any last-minute MatchError inspections."""
+        for match in sorted(matches):
+            if not isinstance(match.rule, TransformMixin):
+                continue
+            match.rule.transform(match, file, data)
```

Comment:
Though this is simple enough to inline in `run()`, it will get more logic in follow-up PRs.